r/tools/linuxlibs.R

```diff
@@ -0,0 +1,131 @@
+
+identify_os <- function(os = Sys.getenv("ARROW_BINARY_DISTRO")) {
+  # Function to figure out which flavor of binary we should download.
+  # By default, it will try to discover from the OS what distro-version we're on
+  # but you can override this by setting the env var ARROW_BINARY_DISTRO to:
+  # * `FALSE` (not case-sensitive), which tells the script not to download a binary,
+  # * some other string, presumably a related "distro-version" that has binaries
+  #   built that work for your OS
+  if (nzchar(os)) {
+    if (identical(tolower(os), "false")) {
+      # Env var says not to download a binary
+      return(NULL)
+    } else {
+      # Env var provided an os-version to use--maybe you're on Ubuntu 18.10 but
+      # we only build for 18.04 and that's fine--so use what the user set
+      return(os)
+    }
+  }
+
+  has_lsb <- system("which lsb_release", ignore.stdout = TRUE) == 0
+  if (has_lsb) {
+    distro <- tolower(system("lsb_release -is", intern = TRUE))
+    os_version <- system("lsb_release -rs", intern = TRUE)
+    # In the future, we may be able to do some mapping of distro-versions to
+    # versions we built for, since there's no way we'll build for everything.
+    os <- paste0(distro, "-", os_version)
+  } else if (file.exists("/etc/system-release")) {
+    # Something like "CentOS Linux release 7.7.1908 (Core)"
+    system_release <- tolower(head(readLines("/etc/system-release"), 1))
+    # Extract from that the distro and the major version number
+    os <- sub("^([a-z]+) .*([0-9]+).*$", "\\1-\\2", system_release)
+  } else {
+    cat("*** Unable to identify current OS/version\n")
+    os <- NULL
+  }
+  return(os)
+}
+
+if (!file.exists(paste0(dst_dir, "/include/arrow/api.h"))) {
+  # If we're working in a local checkout and have already built the libs, we
+  # don't need to do anything. Otherwise,
+  if (length(args) > 1) {
+    # Arg 2 would be the path/to/lib.zip
+    # TODO: do we need this option?
+    localfile <- args[2]
+    if (file.exists(localfile)) {
+      cat(sprintf("*** Using LOCAL_LIBARROW %s\n", localfile))
+      file.copy(localfile, "lib.zip")
+    } else {
+      cat(sprintf("*** LOCAL_LIBARROW %s does not exist\n", localfile))
+    }
+  } else if (download_libs) {
+    base_url <- "https://dl.bintray.com/ursa-labs/arrow-r/libarrow/"
+    # Determine (or take) distro
+    os <- identify_os()
+    if (!is.null(os)) {
+      # Download it, if allowed
+      binary_url <- paste0(base_url, os, "/arrow-", VERSION, ".zip")
+      try(
+        download.file(binary_url, "lib.zip", quiet = quietly),
+        silent = quietly
+      )
+    }
+    if (file.exists("lib.zip")) {
+      cat(sprintf("*** Successfully retrieved C++ binaries for %s\n", os))
+    } else {
+      if (!is.null(os)) {
+        cat(sprintf("*** No C++ binaries found for %s\n", os))
+      }
+      # Try to build C++ from source
+      tf1 <- tempfile()
+      source_url <- paste0(base_url, "src/arrow-", VERSION, ".zip")
```

Comment:
Yeah I thought about that, wondered if you'd ask ;) I think I'll still have it check bintray first, in case there's some amendment we have to make after the official release (we have, at least for the R package but not the C++ code, for each of the last couple of releases in order to satisfy some CRAN check). But I won't put one on bintray for an official release if we can use the official source.

Reply:
I see.